[gohqingkhang] iP

gohqingkhang opened the pull request:
# Delta ChatBot

> “Nothing is particularly hard when you divide it into small jobs.” – Henry Ford ([source](https://dansilvestre.com/productivity-quotes/))

Delta ChatBot frees your mind of having to remember things you need to do. It's,

- text-based
- ~~easy~~ _**SUPER EASY**_ to learn

All you need to do is,

1. download the jar file from [here](https://github.com/gohqingkhang/ip/releases/tag/v0.1)
2. open cmd line and run this file using java -jar Delta.jar
3. add your tasks
4. let Delta ChatBot manage tasks for you! 🤩 

Best of all... It is **FREE!!!**

Features include:

- [x] managing tasks
- [ ] sending reminders

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

gohqingkhang committed:
Refactored name of chatbot from Duke to Delta

gohqingkhang committed:
Add 2 static methods to greet and say bye
Change format of greeting and bye message

gohqingkhang committed:
Import Scanner and create Scanner object

gohqingkhang committed:
Add static method echo

gohqingkhang committed:
Add implementation for echo

gohqingkhang committed:
Add break in logic to end chatbot after bye

gohqingkhang committed:
Add tab for all messages

gohqingkhang committed:
Add arraylist attribute to store user inputs

gohqingkhang committed:
Add static method add to add user input onto list

gohqingkhang committed:
Edit static method echo to print all elements in list

gohqingkhang committed:
Edit logic in main to add implementation for add and echo methods

gohqingkhang committed:
Add Scanner close at end of main

gohqingkhang committed:
Add new Task class

gohqingkhang committed:
Add attrbutes, constructor and methods for Task class

gohqingkhang committed:
Add method to mark task as not done

gohqingkhang committed:
Change method name from echo to printTasks

gohqingkhang committed:
Change method name from add to addTask

gohqingkhang committed:
Add getter method to get description of task

gohqingkhang committed:
Change list to take in type Task instead of String
Edit method addTask to add Task objects
Edit main for type Task

gohqingkhang committed:
Edit method printTasks to print in new format

gohqingkhang committed:
Add method markTask to mark task as done

gohqingkhang committed:
Add method unmarkTask to mark task as not done

gohqingkhang committed:
Edit main to add logic for mark and unmark tasks

gohqingkhang committed:
Add newline for markTask and unmarkTask methods

gohqingkhang committed:
Change indent for markTask and unmarkTask methods

gohqingkhang committed:
Add toString method for Task to print format of Task

gohqingkhang committed:
Create Todo class as subclass of Task

gohqingkhang committed:
Create Deadline class as subclass of Task

gohqingkhang committed:
Create Event class as subclass of Task

gohqingkhang committed:
Add spacing in toString method of Deadline

gohqingkhang committed:
Add logic in main to handle Todos, Deadlines and Events

gohqingkhang committed:
Edit format for method addTask

gohqingkhang committed:
Add tests in input.txt and EXPECTED.TXT to test automated testing

gohqingkhang committed:
removed line to fix formatting problem

gohqingkhang committed:
Create DeltaException class as subclass of Exception
Add try and catch block in main for error handling

gohqingkhang committed:
Split DeltaException and StringIndexOutOfBoundsException

gohqingkhang committed:
Add logic to catch empty Todo, Deadline and Event
Add annotations

gohqingkhang committed:
Edit logic in main to catch empty task

gohqingkhang committed:
Add static method printError to format the printing of errors

gohqingkhang committed:
Change error message for unknown action in main

gohqingkhang committed:
Add logic in main to catch deadline format error

gohqingkhang committed:
Add logic in main to catch event format error

gohqingkhang committed:
Change format for print messages in main

gohqingkhang committed:
Add logic to ignore case sensitivity

gohqingkhang committed:
Edit method printTasks to catch if list is empty
Add logic in main to catch format error for Mark and Unmark Tasks

gohqingkhang committed:
Edit markTask and unmarkTask methods to catch out of range error and double marking error
Change message in addTask method

gohqingkhang committed:
Add logic in main to catch invalid index for mark and unmark tasks

gohqingkhang committed:
Remove redundant initialization of variable

gohqingkhang committed:
Add new static method deleteTask to delete task from list
Add error handling if task index not in range

gohqingkhang committed:
Add error handling in methods markTask, unmarkTask and deleteTask to catch if list is empty

gohqingkhang committed:
Add logic in main to implement deletion of task
Add error handling if invalid index given

gohqingkhang committed:
Edit format for message for deleteTask method

gohqingkhang committed:
Add new test cases in automated testing

gohqingkhang committed:
Edit syntax in code

gohqingkhang committed:
Add comments for all classes and methods
Clean up logic in main

gohqingkhang committed:
Add comment for DeltaException class

gohqingkhang committed:
Add saveTasks method in main to save tasks in ./data/DeltaList.txt
Create enum TaskType fix Tasks created as Todo, Deadline or Event only
Add saveDetails method in task class and subclasses to convert task details into format for saving

gohqingkhang committed:
Add comments for new methods

gohqingkhang committed:
Add logic in main to manually save tasks

gohqingkhang committed:
Create loadTasks method to load up list from save file

gohqingkhang committed:
Add comments to new method loadTasks

gohqingkhang committed:
Catch error when no previous session

gohqingkhang committed:
Merge branch 'branch-Level-7'

gohqingkhang committed:
Create formatDateTime method to check format of user input date/time and return proper format to be printed
Add overloaded constructor for Deadline task to store date/time as LocalDateTime attribute

gohqingkhang committed:
Add comments for new method and constructor added

gohqingkhang committed:
Add overloaded constructor for Event task to store date/time as LocalDateTime attribute

gohqingkhang committed:
Add error checking for dates before current date/time and end must be after start date/time

gohqingkhang committed:
Merge branch 'branch-Level-8'

gohqingkhang committed:
Create Ui class

gohqingkhang committed:
Add readCommand method to take in user input
Edit showLine method to add line divider string
Add showWelcome method to print welcome message
Add showError method to print error message

gohqingkhang committed:
Create Storage class

gohqingkhang committed:
Add load method to load save file contents into an arraylist to be passed to tasklist
Add save method that takes in a TaskList to get tasks to be stored in save list

gohqingkhang committed:
Create Parser class

gohqingkhang committed:
Move main while loop contents from Delta.java main method into Parser.java parse method
Move formatDateTime from Delta.java to Parser.java

gohqingkhang committed:
Create Command class

gohqingkhang committed:
Add abstract method isExit that returns if command is ExitCommand
Add abstract method execute that executes action of current command

gohqingkhang committed:
Create TaskList class

gohqingkhang committed:
Add overloaded constructor to create empty TaskList or loaded TaskList with contents given by ArrayList tasks
Add getter method to return tasks in TaskList

gohqingkhang committed:
Create ExitCommand class as subclass of Command abstract class

gohqingkhang committed:
Command.java: Edit execute method in abstract class Command to take in parameters
ExitCommand.java: Add isExit method to return true to exit chatbot
ExitCommand.java: Add execute method to call ui to print exit message
Ui.java: Add showExit method to print exit message

gohqingkhang committed:
Create AddCommand class as subclass of Command abstract class

gohqingkhang committed:
AddCommand.java: Add isExit method to return false as not ExitCommand
AddCommand.java: Add execute method to call TaskList to add task into list and call ui to print add task message
TaskList.java: Add getSize method to return number of tasks in TaskList
TaskList.java: Add addTask method to add task to list

gohqingkhang committed:
Create DeleteCommand class as subclass of Command abstract class

gohqingkhang committed:
DeleteCommand.java: Add isExit method to return false as is not ExitCommand
DeleteCommand.java: Add execute method to call TaskList to delete task and ui to print delete task message
TaskList.java: Add deleteTask method to delete task from list with given index - 1

gohqingkhang committed:
Create MarkCommand class as subclass of Command abstract class

gohqingkhang committed:
TaskList.java: Add markTask method to mark task in list as done
MarkCommand.java: Add isExit method to return false as is not ExitCommand
MarkCommand.java: Add execute method to call TaskList to mark task as done and call ui to print mark task message
Edit execute method in AddCommand.java, DeleteCommand.java and MarkCommand.java to save TaskList in storage

gohqingkhang committed:
Create UnmarkCommand class as subclass of Command abstract class

gohqingkhang committed:
TaskList.java: Add unmarkTask method to mark task in list as not done
UnmarkCommand.java: Add isExit method to return false as is not ExitCommand
UnmarkCommand.java: Add execute method to call TaskList to mark task as not done and call ui to print unmark task message

gohqingkhang committed:
Create PrintCommand class as subclass of Command abstract class

gohqingkhang committed:
Fix bugs and errors
Change test cases and tested to make sure working

gohqingkhang committed:
Add comments for all classes and methods

gohqingkhang committed:
Create CommandType enum to fix the possible type of Commands

gohqingkhang committed:
Merge branch 'branch-A-MoreOOP'

gohqingkhang committed:
Move all classes into packages
Add import statements for all classes and methods used

gohqingkhang committed:
Merge branch 'branch-A-Packages'

gohqingkhang committed:
Merge remote-tracking branch 'origin/add-gradle-support'

gohqingkhang committed:
Set up Gradle
Edit format to abide by checkstyle rules

gohqingkhang committed:
Merge branch 'branch-A-Gradle'

gohqingkhang committed:
Create AddCommandTest.java, DeadlineTest.java and TaskListTest.java to test methods in each class

gohqingkhang committed:
Merge branch 'branch-A-JUnit'

gohqingkhang committed:
Edit import statements to fit checkstyle requirements

gohqingkhang committed:
Create Delta.jar JAR file

gohqingkhang committed:
Merge branch 'branch-A-Jar'

gohqingkhang committed:
Add JavaDoc comments to all classes and methods

gohqingkhang committed:
Create FindCommand class as subclass of Command abstract class
CommandType.java: Add Find as a possible Command type
Parser.java: Add logic to parse Find command

gohqingkhang committed:
Merge branch 'branch-A-JavaDoc'

gohqingkhang committed:
Add comments for classes and methods

gohqingkhang committed:
Merge branch 'branch-A-CodingStandard'

gohqingkhang committed:
Merge branch 'branch-Level-9'

gohqingkhang committed:
Rearrange items in build.gradle

gohqingkhang committed:
Merge branch 'branch-A-CheckStyle'

gohqingkhang committed:
Add dependecies for JavaFX in build.gradle

gohqingkhang committed:
Refactor printing of string to be in a single private method

gohqingkhang committed:
Add files from javafx tutorial to run gui for Delta ChatBot

gohqingkhang committed:
Add method in Delta.java to get Delta response given user input
Edit execute message format in all command classes

gohqingkhang committed:
Add time delay between user input and delta response
Add logic to close chatbot on bye command

gohqingkhang committed:
Add JavaDoc comments for classes and methods

gohqingkhang committed:
Change png files for user and delta

gohqingkhang committed:
Merge branch 'branch-Level-10'

gohqingkhang committed:
Move GUI files into a separate folder

gohqingkhang committed:
Run clean build on gradle

gohqingkhang committed:
Merge branch 'branch-Level-10'

gohqingkhang committed:
Merge branch 'master' of https://github.com/gohqingkhang/ip

leslieyip02 commented on src/main/java/delta/util/Storage.java:
```diff
@@ -0,0 +1,118 @@
+
+        // Create save directory
+        if (!directory.exists()) {
+            boolean directoryCreatedSuccessfully = directory.mkdir();
```
Boolean variables/methods should be named to sound like booleans. This could be renamed as `isCreatedSuccessfully` instead.

leslieyip02 commented on src/main/java/delta/util/Storage.java:
```diff
@@ -0,0 +1,118 @@
+        // Create save file
+        if (!file.exists()) {
+            try {
+                boolean fileCreatedSuccessfully = file.createNewFile();
```
Boolean variables/methods should be named to sound like booleans. This could be renamed as `isCreatedSuccessfully` instead.

leslieyip02 commented on src/test/java/delta/util/TaskListTest.java:
```diff
@@ -0,0 +1,224 @@
+            fail();
+        } catch (Exception e) {
+            assertEquals("OOPS!!! Task is already marked as not done yet.", e.getMessage());
+        }
```
I like that you test multiple scecnarios of invalid input. Perhaps it would be good to separate these into separate tests?

Using the `featureUnderTest_testScenario_expectedBehavior()` format, this could be refactored into

- `unmarkTask_emptyList_exceptionThrown()`
- `unmarkTask_indexOutOfRange_exceptionThrown()`
- `unmarkTask_unmarkedTask_exceptionThrown()`

leslieyip02 commented on src/main/java/delta/util/Parser.java:
```diff
@@ -0,0 +1,244 @@
+                return new AddCommand(new Todo(taskName));
+            // Only task given, no description
+            } else {
+                throw new DeltaException("""
+                        OOPS!!! Description of todo cannot be left blank!
+                        \t Please follow the proper format:
+                        \t * todo [description]""");
```
Just a suggestion, but maybe it would be good to refactor some of this into the constructor of `DeltaException` because this message pattern is repeated a few times throughout `Parser.java`.

leslieyip02 commented on src/main/resources/images/User.png:
Awesome

zlch commented on src/test/java/delta/util/TaskListTest.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @Test
+    public void testGetTask() {
```
Could the test function name be more descriptive to show what situations are being tested?

zlch reviewed (commented):
LGTM, just some nits

gohqingkhang committed:
Change name of boolean variable used

gohqingkhang committed:
Add assertions for subclasses of Command class

Subclasses assume command is executed as intended after execute method.

Checks should be done to ensure that each Command subclass is executing
the method as intended.

Let's add assertions after the main action of each execute method in
all Command subclasses.

Assertions help to check if the action taken by the execute method is
correct as per intended by the specific subclass.

gohqingkhang committed:
Add assertions for Storage and TaskList classes

gohqingkhang committed:
Edit assertions to comply with checkstyle

gohqingkhang committed:
Extract error messages into constants

Certain states in the parse method throw the same error message.

These same error messages cause code duplication, and also results in
untidy code.

Let's extract these error messages from the parse method and store
them as constants instead.

Consolidating the error messages help when changes are needed and
prevent typos.

gohqingkhang committed:
Change if-else construct to switch

gohqingkhang committed:
Remove deep nesting and add guard clauses

Deeply nested logic for deadline and event commands. Happy path unclear
for other commands.

Code looks very messy, hard to read and follow logic.

Let's remove all the deep nesting and push the guard clauses to the top.

Removing the deep nesting makes the logic in the code easier to follow.

gohqingkhang committed:
Add empty lines for clarity in code

gohqingkhang committed:
Reduce level of abstraction in parse method

All command checking and handling done in a single parse method.

Having an extremely long method makes code harder to be read and makes
finding bugs harder.

Let's extract all the different commands into separate private methods.

Reducing the level of abstraction in the parse method and separating
out the different commands makes the overall code easier to read and
logic easier to follow.

gohqingkhang committed:
Merge pull request #1 from gohqingkhang/branch-A-Assertions

Add assertions in code

gohqingkhang committed:
Merge branch 'master' of https://github.com/gohqingkhang/ip into branch-A-CodeQuality

gohqingkhang committed:
Merge pull request #2 from gohqingkhang/branch-A-CodeQuality

Edit Code Quality

gohqingkhang committed:
Create workflows

gohqingkhang committed:
Delete .github/workflows

gohqingkhang committed:
Create gradle.yml

gohqingkhang committed:
Comment out IO redirection lines

gohqingkhang committed:
Uncomment IO redirection for Windows

gohqingkhang committed:
Comment out Windows IO redirection

gohqingkhang committed:
Add logic in Parser to handle edit command
Create EditCommand class as subclass of Command class
Add editTask method in TaskList to handle editing of tasks
Add getters methods in Deadline and Event class to get unformatted time

gohqingkhang committed:
Add JavaDoc comments for all new classes and methods
Edit code to follow checkstyle standard

gohqingkhang committed:
Merge pull request #3 from gohqingkhang/branch-C-Update

Add c update functionality

gohqingkhang committed:
Add files via upload

gohqingkhang committed:
Delete docs/Ui.png

gohqingkhang committed:
Add files via upload

gohqingkhang committed:
Update README.md to follow Delta user guide

gohqingkhang committed:
Update README.md to add local links to each section

gohqingkhang committed:
Merge pull request #6 from gohqingkhang/branch-A-UserGuide

Branch a user guide

gohqingkhang committed:
Split markTask method testing into individual tests

gohqingkhang committed:
Split unmarkTask method testing into individual tests

gohqingkhang committed:
Split deleteTask method testing into individual tests

gohqingkhang committed:
Add title for app window

gohqingkhang committed:
Edit error message for index out of range error

gohqingkhang committed:
Replace Ui png for user guide
